### PR TITLE
Refactored Case

### PR DESCRIPTION
**Pull request by normanyahq**
Mentioned in #116 

Also, refactored nullif, coalesce, and unary_minus.

### CODE REVIEW

File: src/backend/expression/case_expression.h [hunk 1]
```diff
@@ -9,69 +9,74 @@
...
-
+#include <utility>
 #include "backend/expression/abstract_expression.h"
 #include "backend/expression/expression_util.h"
 
 namespace peloton {
 namespace expression {
 
+// the first part is condition, and the second part is expression
+typedef std::pair<AbstractExpression *, AbstractExpression *> WhenClause;
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: Can you move this `typedef` into `CaseExpression`

File: src/backend/expression/case_expression.h [hunk 2]
```diff
@@ -9,69 +9,74 @@
...
+
 class CaseExpression : public AbstractExpression {
  public:
-  CaseExpression(ValueType vt, const std::vector<AbstractExpression *> &clauses,
-                 AbstractExpression *default_result)
+  CaseExpression(ValueType vt, const std::vector<WhenClause *> &clauses,
+                 WhenClause *default_clause)
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: With the modification to the typedef above, use `std::move` to move the vector parameter's contents.  Also pass in an r-value `unique_ptr&&` for the default clause.

File: src/backend/expression/case_expression.h [hunk 3]
```diff
@@ -9,69 +9,74 @@
...
       : AbstractExpression(EXPRESSION_TYPE_OPERATOR_CASE_EXPR),
         clauses(clauses),
-        default_result(default_result),
+        default_clause(default_clause),
         case_type(vt) {}
 
   ~CaseExpression() {
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: You won't need a destructor with the changes above.

**Comment by pmenon** [comment]
Thanks for the quick turnaround! 

It looks good overall, but I've added a few comments to address.  Try to use `unique_ptr` whenever you see a `new` anywhere in the code.  Apply this technique to any code you're modifying (in this case, CASE, COALESCE and NULLIF expressions too).

Try to make the above changes, and fix the build issue that's come up

**Comment by normanyahq** [comment]
OK... I forgot to modify unit test.
btw, I can't make check on my machine with gcc-5.2.1

**Comment by normanyahq** [comment]
Done. I will add unit tests for nullif, coalesce later.

File: src/backend/expression/case_expression.h [hunk 4]
```diff
@@ -18,60 +20,55 @@ namespace expression {
 
 class CaseExpression : public AbstractExpression {
  public:
-  CaseExpression(ValueType vt, const std::vector<AbstractExpression *> &clauses,
-                 AbstractExpression *default_result)
+  // the first part is condition, and the second part is expression
+  typedef std::unique_ptr<AbstractExpression> AbstractExprPtr;
+  typedef std::pair<AbstractExprPtr, std::unique_ptr<AbstractExpression>>
+      WhenClause;
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: Be consistent with your typedef usage.

File: src/backend/expression/case_expression.h [hunk 5]
```diff
@@ -18,60 +20,55 @@ namespace expression {
...
-      delete clause;
-      clause = nullptr;
-    }
-    if (default_result != nullptr) {
-      delete default_result;
-      default_result = nullptr;
+        default_clause(std::move(default_clause)),
+        case_type(vt) {
+    for (auto &clause : t_clauses) {
+      clauses.push_back(std::move(clause));
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: You should be able to construct `clauses` as `std::move(t_clauses)`

**Comment by pmenon** [comment]
I've added a few minor edits to make.  Also, since we no longer need `OperatorCaseWhenExpression`, can you remove that?

File: src/backend/expression/case_expression.h [hunk 6]
```diff
@@ -9,7 +9,9 @@
 // Copyright (c) 2015-16, Carnegie Mellon University Database Group
 //
 //===----------------------------------------------------------------------===//
-
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: Missing an include guard here.  Just pop in `#pragma once`

File: src/backend/expression/case_expression.h [hunk 7]
```diff
@@ -18,60 +20,55 @@ namespace expression {
...
   }
 
  private:
   // Case expression clauses
-  std::vector<AbstractExpression *> clauses;
+  std::vector<WhenClausePtr> clauses;
 
   // Fallback case result expression
-  AbstractExpression *default_result;
+  WhenClausePtr default_clause;
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: Use the member variable naming convention by putting an underscore as a suffix. Do `clauses_` as an example.

**Comment by pmenon** [comment]
I've added a few minor edits to make. Also, since we no longer need `OperatorCaseWhenExpression`, can you remove that?

File: src/backend/expression/case_expression.h [hunk 8]
```diff
@@ -9,6 +9,7 @@
 // Copyright (c) 2015-16, Carnegie Mellon University Database Group
 //
 //===----------------------------------------------------------------------===//
+#pragma once
```
Comment (pmenon) [inline src/backend/expression/case_expression.h]: Just add some spacing between the comment, the guard and the first include.

**Comment by pmenon** [comment]
This is awesome. Just one final formatting comment and we can merge this in.

**Comment by normanyahq** [comment]
Done. Thank you for all the suggestions. I learned a lot.

**Pull request merged**